plugins/labormanager/joblabormapper.cpp

```diff
@@ -313,14 +311,27 @@ class jlfunc_construct_bld : public jlfunc
                 return workshop_build_labor[ws->type];
         }
         break;
+        case df::building_type::Furnace:
+        {
+            df::building_furnacest* frn = (df::building_furnacest*) bld;
```

Comment:
Maybe `virtual_cast` would be safer?

Reply:
Probably. There are a ton of "potentially unsafe" casts in this code. If I "fix" this one, I should probably fix the one that's like ten lines above it in the code for custom workshops, too...

Reply:
Fair, I didn't expand the diff here. `virtual_cast` would check the actual type, but in any case, this should only ever break due to a structures issue (e.g. a new building_type added somewhere in the middle), which would be pretty easy to catch since it would break a lot of other things too.

Reply:
Also, `virtual_cast` isn't really safer here. All `virtual_cast` is is a wrapper that either casts the pointer to the type or returns null. The code still has to do a null check. But (with current DF structures, at least) we're guaranteed that if `bld->getType()` is Furnace, then `bld` is a `building_furnacest`.

Reply:
virtual_cast also checks DF's type information (not sure if that's what you meant; it's basically dynamic_cast, but we can't use that for DF-defined types). That's based on the `<vtable-address>` entries in symbols.xml, which are a bit more reliable than our enum definitions. But in order for this code to fail, a new building subclass would need to be added whose `getType()` returns what was previously `building_type::Furnace`, and our  `building_type` would need to be outdated, which should really only be possible soon after a DF release. So I agree, it's not really worth changing.